p2p/src/simulated/bandwidth.rs

```diff
@@ -27,275 +30,360 @@ pub enum Rate {
-fn compute_rates(active: &[usize], resources: &[Resource], rates: &mut [Option<Ratio>]) {
-    // If there are no active flows, there is nothing to allocate.
-    if active.is_empty() {
-        return;
-    }
+    /// Register each flow with the constrained resources it depends on. Resources without limits are ignored.
+    fn register<E, I>(&mut self, egress_limit: &mut E, ingress_limit: &mut I)
+    where
+        E: FnMut(&P) -> Option<u128>,
+        I: FnMut(&P) -> Option<u128>,
+    {
+        for (flow_idx, flow) in self.flows.iter().enumerate() {
+            let mut state = State::new();
+
+            // Register the flow with its egress resource if the sender is bandwidth-limited.
+            if let Some(resource_idx) = self.constrain(
+                Constraint::Egress(flow.origin.clone()),
+                egress_limit(&flow.origin),
+            ) {
+                self.attach(resource_idx, flow_idx, &mut state);
+            }
 
-    // Track how much capacity each resource still has to hand out.
-    let mut remaining: Vec<Ratio> = resources.iter().map(|res| res.capacity.clone()).collect();
+            // Only track ingress when the recipient actually needs to receive the bytes.
+            if flow.delivered {
+                // Register the flow with its ingress resource if the recipient is bandwidth-limited.
+                if let Some(resource_idx) = self.constrain(
+                    Constraint::Ingress(flow.recipient.clone()),
+                    ingress_limit(&flow.recipient),
+                ) {
+                    self.attach(resource_idx, flow_idx, &mut state);
+                }
+            }
 
-    // `unfrozen[idx] == true` means the flow is still participating in progressive filling.
-    let mut unfrozen = vec![false; rates.len()];
-    for &idx in active {
-        unfrozen[idx] = true;
+            // The flow participates in progressive filling until one of its constraints saturates.
+            if state.limited {
+                state.active = true;
+                self.active += 1;
+            }
+            self.states.push(state);
+        }
     }
 
-    // Continue allocating until every flow is either unlimited or fully constrained.
-    let mut active_left = active.len();
-    while active_left > 0 {
-        // Find the resource with the smallest remaining capacity.
-        let mut limiting = Vec::new();
-        let mut min_delta: Option<Ratio> = None;
-        for (res_idx, resource) in resources.iter().enumerate() {
-            // Count how many still-active flows are drawing from this resource.
-            let users: u128 = resource
-                .members
-                .iter()
-                .filter(|&&flow_idx| unfrozen[flow_idx])
-                .count() as u128;
-            if users == 0 {
+    /// Finalize all flows that rely on a saturated resource.
+    fn finalize(&mut self, res_idx: usize) {
```

Comment:
We can merge finalize_all <> finalize to avoid confusion on which should be used (finalize was never called directly)